[nix] patch realm as a nodePackage use it from nix store, add to default nix-shell buildDependencies

hlolli opened the pull request:
### Summary:
Realm, which is called trough the ubuntu server js file, needs to be compiled in nix environment if the environment is to be pure. After this commit, it should be possible to start the shell with `nix-shell --pure`, I'd call this a step forward but still there's some things to be done. Preferably the ubuntuserver.js file should be generated within the nix-shell, because we don't want to use the local node_modules, rather the nix-store realm.

The hostname has been merged since I made my previous PR so I'll close that, but I realized I made a mistake, kinda, the problem is, that the nix-shell isn't provided with parameters, to merge the buildInputs from desktop or mobile, which is given a if then else statement in status-react/default.nix. I'd suggest just importing all the buildInputs from both mobile and desktop into the root environment, or alternatively, make 3 ways of starting the shell, one for desktop and one for mobile and third for both?

I'd want to continue on this and make a wrapped js file that starts the dev server, and encourage developers of the nix feature to use `nix-shell --pure` to make sure that it's possible to avoid the infamous "works on my computer", that's what nix can actually solve ;=)

[edit] also want to add that the global node modules has now one default.nix file, I did this to ensure that all global node module overrides are imported everywhere the same. I noticed that the genrated node2nix files were imported twice. Notebene, the latest master branch node2nix (1.6.1) must be used, some nested dep will fail when using node2nix 1.6.0 release.

### Review notes (optional):

### Testing notes (optional):

#### Platforms (optional)
- Linux

for the realm-core http call, a macos endpoint needs to be added

#### Areas that maybe impacted (optional)
none

status: wip <!-- Can be ready or wip -->

(wip but this branch is a wip so I'd be happy to see this merged).

status-im-auto commented:
### Jenkins Builds
| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :x: | 28b88ca4 | [#1](https://ci.status.im/job/status-react/job/prs/job/android-e2e/job/PR-7759/1/) | 2019-03-19 01:51:35 | ~35 sec | `android-e2e` | [:page_facing_up: log](https://ci.status.im/job/status-react/job/prs/job/android-e2e/job/PR-7759/1/consoleText) |
| :x: | 28b88ca4 | [#1](https://ci.status.im/job/status-react/job/prs/job/android/job/PR-7759/1/) | 2019-03-19 01:51:35 | ~33 sec | `android` | [:page_facing_up: log](https://ci.status.im/job/status-react/job/prs/job/android/job/PR-7759/1/consoleText) |

status-im-auto commented:
### Jenkins Builds
| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :x: | 28b88ca4 | [#1](https://ci.status.im/job/status-react/job/prs/job/android-e2e/job/PR-7759/1/) | 2019-03-19 01:51:35 | ~35 sec | `android-e2e` | [:page_facing_up: log](https://ci.status.im/job/status-react/job/prs/job/android-e2e/job/PR-7759/1/consoleText) |
| :x: | 28b88ca4 | [#1](https://ci.status.im/job/status-react/job/prs/job/android/job/PR-7759/1/) | 2019-03-19 01:51:35 | ~33 sec | `android` | [:page_facing_up: log](https://ci.status.im/job/status-react/job/prs/job/android/job/PR-7759/1/consoleText) |
| :x: | 28b88ca4 | [#1](https://ci.status.im/job/status-react/job/prs/job/ios/job/PR-7759/1/) | 2019-03-19 01:51:35 | ~30 sec | `ios` | [:page_facing_up: log](https://ci.status.im/job/status-react/job/prs/job/ios/job/PR-7759/1/consoleText) |

status-im-auto commented:
### Jenkins Builds
| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :x: | 28b88ca4 | [#1](https://ci.status.im/job/status-react/job/prs/job/android-e2e/job/PR-7759/1/) | 2019-03-19 01:51:35 | ~35 sec | `android-e2e` | [:page_facing_up: log](https://ci.status.im/job/status-react/job/prs/job/android-e2e/job/PR-7759/1/consoleText) |

pedropombeiro commented on .TOOLVERSIONS:
```diff
@@ -1,4 +1,6 @@
 android-sdk-build-tools;28.0.1;
 android-sdk-platform;android-27;
 android-sdk;4333796;aa190cfd7299cd6a1c687355bb2764e4
+node;10.12.0;
```
Couple of things:
- in this branch, .TOOLVERSIONS will only contain tools which are not instaled through Nix
- you mention node 10.12.0, but the `yarn` expression depends on `node` 8.x, I think that's why the macOS build is failing

hlolli replied:
ok, since you were callling toolversions from the node2nix script, I thought it got missing. So the nix branch should be compiled against node 8.x for now then?

hlolli replied:
what yarn expression are you refering to?

pedropombeiro replied:
> ok, since you were callling toolversions from the node2nix script, I thought it got missing. So the nix branch should be compiled against node 8.x for now then?

oh right, that was a mistake, I shouldn't have removed `react_native_cli` from `.TOOLVERSIONS`. I'll fixup the commit.

pedropombeiro replied:
> what yarn expression are you refering to?

`yarn` here is based on 8.x: https://github.com/status-im/status-react/blob/feature%2Fnix/default.nix#L34. Initially I was using node 10.x until I realized that it was causing builds to fail due a mismatch in node versions.

It's probably possible to override the yarn package to use the nodejs package we want, but there wasn't really a need to be on the latest nodejs version, so I kept it simple for this initial PR.

pedropombeiro commented on scripts/lib/setup/nix/global-node-packages/default.nix:
```diff
@@ -0,0 +1,24 @@
+{ pkgs, nodejs }:
+
+let nodePackages = import ./output {inherit pkgs; inherit nodejs;};
```
```suggestion
let nodePackages = import ./output { inherit pkgs; inherit nodejs; };
```

hlolli replied:
done

PombeirP commented:
> After this commit, it should be possible to start the shell with `nix-shell --pure`

Do you get an error in `feature/nix` when starting with `nix-shell --pure`? I have to admit I haven't looked into that option yet, as I've mostly been focused on getting a minimal working version for all platforms.

PombeirP commented:
> but I realized I made a mistake, kinda, the problem is, that the nix-shell isn't provided with parameters, to merge the buildInputs from desktop or mobile, which is given a if then else statement in status-react/default.nix. I'd suggest just importing all the buildInputs from both mobile and desktop into the root environment, or alternatively, make 3 ways of starting the shell, one for desktop and one for mobile and third for both?

I'm not sure I follow, the current `/default.nix` will import everything if you don't pass arguments to `nix-shell`, and you can also pass `--argstr target-os android` if you only want to set up an environment for Android.

pedropombeiro reviewed (commented):
Thanks for the PR! Didn't have yet enough time to go over all of it, but left a few comments/questions.

hlolli commented:
> I'm not sure I follow, the current /default.nix will import everything if you don't pass arguments to nix-shell, and you can also pass --argstr target-os android if you only want to set up an environment for Android.

I'm refering to these lines
https://github.com/status-im/status-react/blob/feature/nix/default.nix#L55-L56

You are right, all the dependencies are fetched, but for the current shell, which is based on this expression, we wouldn't have access to the binaries of the nested derivation. Which isn't nececcarily needed. Yes it's possible to add arg like this, but I'm wondering what should the behaviour of `make shell` be?

Otherwise all the binaries should be called from within nix so that (in perfect world) we wouldn't need direct access to any build inputs.

> Do you get an error in feature/nix when starting with nix-shell --pure? I have to admit I haven't looked into that option yet, as I've mostly been focused on getting a minimal working version for all platforms.

Don't remember if I did get an error, with this commit I'm able to run `make prepare-desktop` but this is just a step in the direction of being able to run `make run-deskptop` and `node ubuntu-server.js`. I don't want to make my commit too big in case it won't get merged :)

If you want a version working on all platforms, you should use nix-shell --pure to make sure your global (ubuntu?) environment isn't a factor into the success of the build, otherwise there isn't much use of nix.

PombeirP commented:
If you rebase it should fix the `jenkins/prs/android-e2e` build.

PombeirP commented:
> (wip but this branch is a wip so I'd be happy to see this merged).

The feature/nix branch has already been tested and will likely be merged tomorrow, but would be great to have a follow-up PR focusing on improving the situation on NixOS.

PombeirP commented:
Regarding `hostname`, looks like the version present in nixpkgs isn't going to cut it. Here's what I get on my machine:

``` shell
[nix-shell:~/src/github.com/status-im/status-react]$ hostname -i -v
gethostname()=`pedros-thinkpad-x1'
Resolving `pedros-thinkpad-x1' ...
hostname: No address associated with name
```

The 3.20 version in Ubuntu works just fine and returns multiple IP addresses.

hlolli commented:
I'd say I'm surprised as well to why nixpkgs includes such an old version of hostname :/ maybe there's a deeper reason for it. Also intetutils seems to have not been released for a long time since 1.9.4 http://ftp.gnu.org/gnu/inetutils/ maybe the hostname package isn't the one that should be used here.

But about my PR, there are some duplicates of build deps (ex. ncurses for tcl in desktop and root default.nix) and it isn't as pretty as it can be. But it's a base for me to do more PR's and when the behaviour gets more clear, we can reduce redundancy.

PombeirP commented:
@hlolli the Nix branch has been merged to develop and the script no longer requires `hostname` for our use case.

One thing I wanted to ask you, this PR builds `compiled/realm.node` right? We also have a copy in `node_modules/react` and Desktop uses that when in debug mode (for release builds it is downloading a pre-built image, which I'd like to get rid of). I'm not yet clear how these things get built and used (not very familiar with React/npm development other than basic projects).

Another thing I noticed, looks like in NixOS it is not using our Nix cache at nix-cache.status.im. Not sure why this difference of behavior to other distros, we're passing a config path in `NIX_CONF_DIR`: https://github.com/status-im/status-react/blob/feature%2Fadd-jenkins-docker-image/Makefile#L64. Do you have insight into that?

hlolli commented:
Hi @PombeirP,

yes, I'm sure it has to do with the fact that you import nixpkgs here

https://github.com/status-im/status-react/blob/feature/add-jenkins-docker-image/default.nix#L2

There could be two ways to set the channel to status's, one with env variable, otherwise directly in between the <> brackets. Let me look it up, my guess would be not to change the nix config dir, I think it's only for local root nix file.

hlolli commented:
Maybe you can make a short nix expression file that uses fetchFromGit and returns pkgs and not to use pkgs as argument (it may get automatically provided from the nix-shell command), and just under the {}: arguments do

let pkgs = import ./status-channel.nix {};

and then inherit pkgs like is done to children. I guess we can test it.

hlolli commented:
I notice the feture/nix has been merged into develop. I'll rebase the commit and request to merge to develop, and add `realm.node` for macos. I notice that there's different realm for desktop and mobile, for now this is only for the desktop. In nixos I get linker error for pre-built node binaries, by building it with node-gyp, this will work.

hlolli committed:
[nix] patch realm as a nodePackage for linux and darwin and use it from nix store, add to default nix-shell buildDependencies, regenerate nodePackages with node2nix 1.6.1

hlolli commented:
So I added darwin endpoint for realm, deleted the node 10 statement in toolsversion and default.nix. Hope it's good.

Noticed btw some new error after rebasing, with gcc/g++. Will make another PR.

PombeirP commented:
FYI https://discuss.status.im/t/the-road-to-nix-a-functional-package-manager-to-rule-them-all/1049/67

flexsurfer commented:
@hlolli @PombeirP  is this PR still relevant?

jakubgs commented:
Pedro is on holidays.

PombeirP commented:
I still haven't had time to look into this PR in-depth, hopefully I'll find some time soon.

PombeirP commented:
fwiw, rebased and pushed to origin/nix-patch-realm

PombeirP commented:
@hlolli in case you're still active on this, could you please test the current upstream develop to see if you still have issues starting a nix shell with --pure? I've tried `TARGET_OS=android make shell` on NixOS 19.03 and it worked fine even without this PR (for Android `make shell` uses `--pure`).

hlolli commented:
Yes I'll have a look this week. I'll be hacking with Yenda when he's back
:=)

Otherwise I've been iOS oriented in other projects I have going on lately,
was hoping svandeburg would be working more on gradle for nixos.

On Mon, 6 May 2019 at 16:51, Pedro Pombeiro <notifications@github.com>
wrote:

> @hlolli <https://github.com/hlolli> in case you're still active on this,
> could you please test the current upstream develop to see if you still have
> issues starting a nix shell with --pure? I've tried TARGET_OS=android
> make shell on NixOS 19.03 and it worked fine even without this PR (for
> Android make shell uses --pure).
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/status-im/status-react/pull/7759#issuecomment-489649587>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABOLDAU2KKCS54JTJOG5743PUBAWRANCNFSM4G7MODQQ>
> .
>

hlolli commented:
Everything works on the develop branch, until when I want to run it on the emulator

```
➜  status-react git:(develop) ✗ make run-android                                                                              ~/forks/status-react
react-native run-android --appIdSuffix debug
Finished!
yarn install v1.13.0
[1/4] Resolving packages...
success Already up-to-date.
Done in 0.45s.
Finished!
Configuring pure Nix shell for target 'android'...
info JS server already running.
info Building and installing the app on the device (cd android && ./gradlew app:installDebug)...
Starting a Gradle Daemon, 3 stopped Daemons could not be reused, use --status for details

> Configure project :app
Reading env from: .env
Build number file missing, using: 9999
WARNING: Configuration 'compile' is obsolete and has been replaced with 'implementation' and 'api'.
It will be removed at the end of 2018. For more information see: http://d.android.com/r/tools/update-dependency-configurations.html
registerResGeneratingTask is deprecated, use registerGeneratedResFolders(FileCollection)
registerResGeneratingTask is deprecated, use registerGeneratedResFolders(FileCollection)
registerResGeneratingTask is deprecated, use registerGeneratedResFolders(FileCollection)

> Configure project :react-native-background-timer
WARNING: Configuration 'compile' is obsolete and has been replaced with 'implementation' and 'api'.
It will be removed at the end of 2018. For more information see: http://d.android.com/r/tools/update-dependency-configurations.html

> Configure project :react-native-config
WARNING: Configuration 'compile' is obsolete and has been replaced with 'implementation' and 'api'.
It will be removed at the end of 2018. For more information see: http://d.android.com/r/tools/update-dependency-configurations.html

> Configure project :react-native-dialogs
WARNING: Configuration 'compile' is obsolete and has been replaced with 'implementation' and 'api'.
It will be removed at the end of 2018. For more information see: http://d.android.com/r/tools/update-dependency-configurations.html
WARNING: Configuration 'testCompile' is obsolete and has been replaced with 'testImplementation'.
It will be removed at the end of 2018. For more information see: http://d.android.com/r/tools/update-dependency-configurations.html

> Configure project :react-native-fs
WARNING: Configuration 'compile' is obsolete and has been replaced with 'implementation' and 'api'.
It will be removed at the end of 2018. For more information see: http://d.android.com/r/tools/update-dependency-configurations.html

> Configure project :react-native-http-bridge
WARNING: Configuration 'compile' is obsolete and has been replaced with 'implementation' and 'api'.
It will be removed at the end of 2018. For more information see: http://d.android.com/r/tools/update-dependency-configurations.html

> Configure project :react-native-image-crop-picker
WARNING: Configuration 'compile' is obsolete and has been replaced with 'implementation' and 'api'.
It will be removed at the end of 2018. For more information see: http://d.android.com/r/tools/update-dependency-configurations.html

> Configure project :react-native-image-resizer
WARNING: Configuration 'compile' is obsolete and has been replaced with 'implementation' and 'api'.
It will be removed at the end of 2018. For more information see: http://d.android.com/r/tools/update-dependency-configurations.html

> Configure project :react-native-keychain
WARNING: Configuration 'compile' is obsolete and has been replaced with 'implementation' and 'api'.
It will be removed at the end of 2018. For more information see: http://d.android.com/r/tools/update-dependency-configurations.html

> Configure project :react-native-securerandom
WARNING: Configuration 'compile' is obsolete and has been replaced with 'implementation' and 'api'.
It will be removed at the end of 2018. For more information see: http://d.android.com/r/tools/update-dependency-configurations.html

> Configure project :react-native-splash-screen
WARNING: Configuration 'compile' is obsolete and has been replaced with 'implementation' and 'api'.
It will be removed at the end of 2018. For more information see: http://d.android.com/r/tools/update-dependency-configurations.html
WARNING: Configuration 'testCompile' is obsolete and has been replaced with 'testImplementation'.
It will be removed at the end of 2018. For more information see: http://d.android.com/r/tools/update-dependency-configurations.html

> Configure project :react-native-status
WARNING: Configuration 'compile' is obsolete and has been replaced with 'implementation' and 'api'.
It will be removed at the end of 2018. For more information see: http://d.android.com/r/tools/update-dependency-configurations.html

> Configure project :react-native-webview
:react-native-webview:reactNativeAndroidRoot /home/hlolli/forks/status-react/node_modules/react-native/android

> Configure project :react-native-webview-bridge
WARNING: Configuration 'compile' is obsolete and has been replaced with 'implementation' and 'api'.
It will be removed at the end of 2018. For more information see: http://d.android.com/r/tools/update-dependency-configurations.html

> Task :app:hemroidBuild FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':app:hemroidBuild'.
> A problem occurred starting process 'command 'null/ndk-build''

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 7s
1 actionable task: 1 executed
error Could not install the app on the device, read the error above for details.
Make sure you have an Android emulator running or a device connected and have
set up your Android development environment:
https://facebook.github.io/react-native/docs/getting-started.html
error Command failed: ./gradlew app:installDebug
make: *** [Makefile:144: run-android] Error 1
```

Did you get it running on the emulator on NixOs19.03?

PombeirP commented:
I haven't been testing on NixOS tbh. What I find strange in those logs is the null in `null/ndk-build`

PombeirP commented:
Hey @hlolli. One thing I noticed is that in this PR you're using the `v2.20.1` branch of the status-im fork (which is normally used for desktop), while the mobile platform on `develop` is actually using `realm@^2.24.0`.

hlolli commented:
That mismatch could be because of a node2nix bug. I got the latest one from npm i -g on the latest commit hash (at the time) of svendeburg's node2nix repository.

The version, if I remember correctly, was one of 2 versions for relm being used, they were different for desktop and mobile, I guess I need to rebase and fix this PR if it's still relevant. The goal was to fix only a realm bug that I got at the time. Is realm causing headaches for you? If so, I'll have another look at this.
(somehow the conversation went bit off track to the original bug I was fixing, so I took it for solved)

PombeirP commented:
I tried with latest 1.6.1 installed with npm and still get that. It does sound like a node2nix bug. The reason I'm looking into realm with Nix is that we have a RN desktop node server process which has so far only been generated by hand, and I'm trying to bring it to Nix. This server contains a realm.node process generated by node-gyp and the process is bundled with `pkg` (that part I've already taken care of). In that branch I'm hitting the same sha256 mismatch issue.

hlolli commented:
Ok, we can override that GCDWebServer with our own endpoint and sha. That could be an alternative to replaceing the sha by hand in the generated file. Coulde be overrideAttrs in this case..

hlolli commented:
No sorry, it's actually realm that's getting this.. Let me look better before replying :)

PombeirP commented:
FWIW, I was able to run the Android app on Genymotion emulator with `make release-android` from the rebased branch.

hlolli commented:
When trying to build desktop I get something different

```
$ make release-desktop 
make[1]: Entering directory '/home/hlolli/forks/status-react'
git clean -qdxf -f ./index.desktop.js desktop/ && \
lein prod-build-desktop
Warning: implicit middleware found: cider-nrepl.plugin/middleware 
Please declare all middleware in :middleware as implicit loading is deprecated.
Warning: implicit hook found: lein-githooks.plugin/hooks 
Hooks are deprecated and will be removed in a future version.
Updating hook:  pre-commit
Updating hook:  pre-commit
Updating hook:  pre-commit
Compiling ClojureScript...
Retrieving org/clojure/clojurescript/1.10.516/clojurescript-1.10.516.pom from central
Compiling ["index.desktop.js"] from ["components/src" "react-native/src/cljsjs" "react-native/src/desktop" "src" "env/prod"]...
Successfully compiled ["index.desktop.js"] in 86.93 seconds.
make[1]: Leaving directory '/home/hlolli/forks/status-react'
\033[0;31mQT_PATH environment variable is not defined!\033[0m
make: *** [Makefile:79: release-desktop] Error 1
```

hlolli commented:
I see sorry, I'm updating realm now to 2.24.0....

PombeirP commented:
Hmm, that `QT_PATH environment variable is not defined!` message means that the variable wasn't defined for some reason here: https://github.com/status-im/status-react/blob/nix-patch-realm/nix/desktop/linux/default.nix#L25

PombeirP commented:
I was able to make realm build with the following changes:

``` patch
diff --git a/nix/global-node-packages/default.nix b/nix/global-node-packages/default.nix
index 779aec749..787f3f69e 100644
--- a/nix/global-node-packages/default.nix
+++ b/nix/global-node-packages/default.nix
@@ -1,20 +1,20 @@
 { pkgs, nodejs }:
 
 let nodePackages = import ./output { inherit pkgs; inherit nodejs; };
-    realm-version = "5.12.1";
-    realm-patched-name = "realm-git+https://github.com/status-im/realm-js.git#heads/v2.20.1";
+    realm-version = "5.14.0";
+    realm-patched-name = "realm-git+https://github.com/realm/realm-js.git#v2.26.1";
     realm-core-src = pkgs.fetchurl (
       if builtins.currentSystem == "x86_64-darwin" then {
         url = "https://static.realm.io/downloads/core/realm-core-Release-v${realm-version}-Darwin-devel.tar.gz";
-        sha256 = "05ji1zyskwjj8p5i01kcg7h1cxdjj62fcsp6haf2f65qshp6r44d";
+        sha256 = "1z78h3b27w5j3faqx6zzpicl037x8j5cc9r9q42cbhxrrap5fcak";
       } else {
         url = "https://static.realm.io/downloads/core/realm-core-Release-v${realm-version}-Linux-devel.tar.gz";
-        sha256 = "02pvi28qnvzdv7ghqzf79bxn8id9s7mpp3g2ambxg8jrcrkqfvr1";
+        sha256 = "1lxsyfhqxkb6zkvb88yq7lzdjj5iylhk6x7hkgpmh6x5mkrp0m5h";
       }
     );
     realm-dest-dir = if builtins.currentSystem == "x86_64-darwin" then
-      "$out/lib/node_modules/realm/vendor/realm-darwin-x64" else
-      "$out/lib/node_modules/realm/vendor/realm-linux-x64";
+      "$out/lib/node_modules/realm/compiled/node-v64_darwin_x64/realm.node" else
+      "$out/lib/node_modules/realm/compiled/node-v64_linux_x64/realm.node";
 in nodePackages // {
   ${realm-patched-name} = nodePackages.${realm-patched-name}.override(oldAttrs: {
     buildInputs = oldAttrs.buildInputs ++ [ pkgs.nodePackages.node-pre-gyp ];
@@ -22,7 +22,7 @@ in nodePackages // {
     preRebuild = ''
       # Do not attempt to do any http calls!
       substituteInPlace $out/lib/node_modules/realm/scripts/download-realm.js \
-        --replace "return acquire(requirements, realmDir)" ""
+        --replace "!shouldSkipAcquire(realmDir, requirements, options.force)" "false"
       mkdir -p ${realm-dest-dir}
       tar -xzf ${realm-core-src} -C ${realm-dest-dir}
     '';
diff --git a/nix/global-node-packages/generate-nix.sh b/nix/global-node-packages/generate-nix.sh
index 39945831b..4ba200d5a 100755
--- a/nix/global-node-packages/generate-nix.sh
+++ b/nix/global-node-packages/generate-nix.sh
@@ -12,7 +12,7 @@ react_native_cli_required_version=$($toolversion react_native_cli)
 cat << EOF > $input
 [
     { "react-native-cli": "${react_native_cli_required_version}",
-      "realm": "git+https://github.com/status-im/realm-js.git#heads/v2.20.1" }
+      "realm": "git+https://github.com/realm/realm-js.git#v2.26.1" }
 ]
 EOF
```

Not clear why it wasn't finding `realm-dest-dir` without the change.

PombeirP commented:
Btw, I was thinking that we can remove react-native-cli from that file, since that's installed by yarn anyway. We just have to make sure that `node_modules/.bin` is added to the PATH.

PombeirP commented:
I'm making a commit that'll remove react-native-cli from there and split the expression into desktop and mobile versions. It that all works we can merge this.

hlolli commented:
yes, I already deleted that, I'm still struggling, some nested git sha's not resolving correctly

Commit:
nix: Move debug realm into desktop folder and use RN cli from node_modulesSigned-off-by: Pedro Pombeiro <pombeirp@users.noreply.github.com>

PombeirP commented:
> yes, I already deleted that, I'm still struggling, some nested git sha's not resolving correctly

My way around it was to point to the `.tar.gz` artifact instead of pointing to the repo.

hlolli commented:
I think I've solved this, wait few minutes :)

hlolli commented:
sorry, for spamming (we are not on chat), of course using npm registry makes more sense in this case

hlolli commented:
@PombeirP solved this particular issue. 2 things, 1. by deleting the react-native npm package, I get error for that package is missing when I do make build-x. I only assume it would make sense to keep it there? and 2. the gereate-node-packages.sh is using node version 6? I remember this was the intended version, but somehow node 10x is used when building these files.

hlolli commented:
And also, any chance of getting some Status coins, I don't have a wallet, but if it's a possibility of a small tip, I'd set one up when @yenda is back from swarm-conf

PombeirP commented:
> sorry, for spamming (we are not on chat), of course using npm registry makes more sense in this case

no worries about that :-)

> @PombeirP solved this particular issue. 2 things, 1. by deleting the react-native npm package, I get error for that package is missing when I do make build-x. I only assume it would make sense to keep it there? 

If you look at my commit, that is solved in `shell.nix` by adding to the `PATH`.

> and 2. the gereate-node-packages.sh is using node version 6? I remember this was the intended version, but somehow node 10x is used when building these files.

Oops, need to add the intended version to `.TOOLVERSIONS`.

> And also, any chance of getting some Status coins, I don't have a wallet, but if it's a possibility of a small tip, I'd set one up when @yenda is back from swarm-conf

Definitely! Pinging @StatusSceptre to see what we can do about that.

pedropombeiro commented on nix/global-node-packages/default.nix:
```diff
@@ -0,0 +1,31 @@
+    preRebuild = ''
+      # Do not attempt to do any http calls!
+      # delete the last callback at the bottom of download-realm.js
+      head -265 $out/lib/node_modules/realm/scripts/download-realm.js > \
```
This doesn't look very safe compared to `substituteInPlace`. Why this choice?

``` bash
      substituteInPlace $out/lib/node_modules/realm/scripts/download-realm.js \
        --replace "!shouldSkipAcquire(realmDir, requirements, options.force)" "false"
```

hlolli replied:
yes this isn't safe and could need a change when changing realm versions. I got syntax error for the dot in then with substituteInPlace, even if I returned true.

hlolli replied:
alternatively, find the last expression and use that line number instead of hardcoding?

pedropombeiro replied:
> yes this isn't safe and could need a change when changing realm versions. I got syntax error for the dot in then with substituteInPlace, even if I returned true.

Newbie question: what do you mean with "the dot"? How can I reproduce the failure?

Since we now have 2 separate expressions for mobile and desktop, we could also make a .patch file for each.

PombeirP commented:
> Preferably the ubuntuserver.js file should be generated within the nix-shell, because we don't want to use the local node_modules, rather the nix-store realm.

This work is happening in https://github.com/status-im/status-react/issues/8246, so this branch will be a good base for that.

hlolli commented:
@PombeirP will do

PombeirP commented:
@hlolli regarding this in OP:

> for the realm-core http call, a macos endpoint needs to be added

can you elaborate on what needs to be done?

hlolli commented:
No this must be old, because I already have Darwin and Linux endpoints.

It's force-pushed now.

hlolli commented:
I didn't get any c-compilation on realm, maybe it was cached, hopefully. The dot node realm files should work as is worst case.

PombeirP commented:
> I didn't get any c-compilation on realm (from your branch), maybe it was cached, hopefully. The dot node realm files should work as is worst case.

I thought that was expected since we're placing realm.node in `lib/node_modules/realm/compiled/node-v64_linux_x64/realm.node`. I'm not too familiar with the realm project to be honest.

PombeirP commented:
Tried my latest branch and got this when running `make release-android`:

```
requires-port@1.0.0 /nix/store/0pk2swrhhqf1b073xyrwpz6cr5y0jjiq-node-realm-2.26.1/lib/node_modules/realm/node_modules/requires-port
npm ERR! code E404
npm ERR! 404 Not Found: ini@1.3.5

npm ERR! A complete log of this run can be found in:
npm ERR!     /private/var/folders/k4/9md4_0b547x90zh9x5pt1d_r0000gn/T/nix-build-node-realm-2.26.1.drv-0/.npm/_logs/2019-05-22T17_47_14_642Z-debug.log
builder for '/nix/store/rbkvna5swb0slgjbxmnk3ndd6lcs3zn4-node-realm-2.26.1.drv' failed with exit code 1
error: build of '/nix/store/rbkvna5swb0slgjbxmnk3ndd6lcs3zn4-node-realm-2.26.1.drv' failed
```

Not sure if it's related to the change of source to the npm registry.

hlolli commented:
It looks to me that you've stopped all npm install for all packages.. let me look

hlolli commented:
This is what I got

```
$ make release-android
make[1]: Entering directory '/home/hlolli/forks/status-react'
lein prod-build-android
Warning: implicit middleware found: cider-nrepl.plugin/middleware 
Please declare all middleware in :middleware as implicit loading is deprecated.
Warning: implicit hook found: lein-githooks.plugin/hooks 
Hooks are deprecated and will be removed in a future version.
Updating hook:  pre-commit
Updating hook:  pre-commit
Updating hook:  pre-commit
Compiling ClojureScript...
Compiling ["index.android.js"] from ["components/src" "react-native/src/cljsjs" "react-native/src/mobile" "src" "env/prod"]...
Compiling ["index.android.js"] failed.
clojure.lang.ExceptionInfo: failed compiling file:/home/hlolli/forks/status-react/src/status_im/utils/js_resources.cljs {:file #object[java.io.File 0xac4d34b "/home/hlolli/forks/status-react/src/status_im/utils/js_resources.cljs"]}
	at clojure.core$ex_info.invokeStatic(core.clj:4739)
	at clojure.core$ex_info.invoke(core.clj:4739)
	at cljs.compiler$compile_file$fn__3702.invoke(compiler.cljc:1562)
	at cljs.compiler$compile_file.invokeStatic(compiler.cljc:1522)
	at cljs.compiler$compile_file.invoke(compiler.cljc:1498)
	at cljs.closure$compile_file.invokeStatic(closure.clj:573)
	at cljs.closure$compile_file.invoke(closure.clj:564)
	at cljs.closure$fn__5124.invokeStatic(closure.clj:653)
	at cljs.closure$fn__5124.invoke(closure.clj:647)
	at cljs.closure$fn__5052$G__5045__5059.invoke(closure.clj:521)
	at cljs.closure$fn__5128.invokeStatic(closure.clj:662)
	at cljs.closure$fn__5128.invoke(closure.clj:647)
	at cljs.closure$fn__5052$G__5045__5059.invoke(closure.clj:521)
	at cljs.closure$compile_sources$iter__5250__5254$fn__5255.invoke(closure.clj:1011)
	at clojure.lang.LazySeq.sval(LazySeq.java:40)
	at clojure.lang.LazySeq.seq(LazySeq.java:49)
	at clojure.lang.Cons.next(Cons.java:39)
	at clojure.lang.RT.next(RT.java:706)
	at clojure.core$next__5108.invokeStatic(core.clj:64)
	at clojure.core$dorun.invokeStatic(core.clj:3134)
	at clojure.core$doall.invokeStatic(core.clj:3140)
	at clojure.core$doall.invoke(core.clj:3140)
	at cljs.closure$compile_sources.invokeStatic(closure.clj:1007)
	at cljs.closure$compile_sources.invoke(closure.clj:996)
	at cljs.closure$build.invokeStatic(closure.clj:2817)
	at cljs.closure$build.invoke(closure.clj:2718)
	at cljs.build.api$build.invokeStatic(api.clj:208)
	at cljs.build.api$build.invoke(api.clj:189)
	at cljs.build.api$build.invokeStatic(api.clj:195)
	at cljs.build.api$build.invoke(api.clj:189)
	at cljsbuild.compiler$compile_cljs$fn__718.invoke(compiler.clj:61)
	at cljsbuild.compiler$compile_cljs.invokeStatic(compiler.clj:60)
	at cljsbuild.compiler$compile_cljs.invoke(compiler.clj:48)
	at cljsbuild.compiler$run_compiler.invokeStatic(compiler.clj:168)
	at cljsbuild.compiler$run_compiler.invoke(compiler.clj:129)
	at user$eval847$iter__895__899$fn__900$fn__926.invoke(form-init3297578457875040857.clj:1)
	at user$eval847$iter__895__899$fn__900.invoke(form-init3297578457875040857.clj:1)
	at clojure.lang.LazySeq.sval(LazySeq.java:40)
	at clojure.lang.LazySeq.seq(LazySeq.java:49)
	at clojure.lang.RT.seq(RT.java:528)
	at clojure.core$seq__5124.invokeStatic(core.clj:137)
	at clojure.core$dorun.invokeStatic(core.clj:3125)
	at clojure.core$doall.invokeStatic(core.clj:3140)
	at clojure.core$doall.invoke(core.clj:3140)
	at user$eval847.invokeStatic(form-init3297578457875040857.clj:1)
	at user$eval847.invoke(form-init3297578457875040857.clj:1)
	at clojure.lang.Compiler.eval(Compiler.java:7062)
	at clojure.lang.Compiler.eval(Compiler.java:7052)
	at clojure.lang.Compiler.load(Compiler.java:7514)
	at clojure.lang.Compiler.loadFile(Compiler.java:7452)
	at clojure.main$load_script.invokeStatic(main.clj:278)
	at clojure.main$init_opt.invokeStatic(main.clj:280)
	at clojure.main$init_opt.invoke(main.clj:280)
	at clojure.main$initialize.invokeStatic(main.clj:311)
	at clojure.main$null_opt.invokeStatic(main.clj:345)
	at clojure.main$null_opt.invoke(main.clj:342)
	at clojure.main$main.invokeStatic(main.clj:424)
	at clojure.main$main.doInvoke(main.clj:387)
	at clojure.lang.RestFn.applyTo(RestFn.java:137)
	at clojure.lang.Var.applyTo(Var.java:702)
	at clojure.main.main(main.java:37)
Caused by: clojure.lang.ExceptionInfo: node_modules/web3/dist/web3.min.js (No such file or directory) at line 13 /home/hlolli/forks/status-react/src/status_im/utils/js_resources.cljs {:file "/home/hlolli/forks/status-react/src/status_im/utils/js_resources.cljs", :line 13, :column 16, :tag :cljs/analysis-error}
	at clojure.core$ex_info.invokeStatic(core.clj:4739)
	at clojure.core$ex_info.invoke(core.clj:4739)
	at cljs.analyzer$error.invokeStatic(analyzer.cljc:697)
	at cljs.analyzer$error.invoke(analyzer.cljc:693)
	at cljs.analyzer$macroexpand_1.invokeStatic(analyzer.cljc:3342)
	at cljs.analyzer$macroexpand_1.invoke(analyzer.cljc:3338)
	at cljs.analyzer$analyze_seq.invokeStatic(analyzer.cljc:3375)
	at cljs.analyzer$analyze_seq.invoke(analyzer.cljc:3355)
	at cljs.analyzer$analyze_form.invokeStatic(analyzer.cljc:3545)
	at cljs.analyzer$analyze_form.invoke(analyzer.cljc:3541)
	at cljs.analyzer$analyze_STAR_.invokeStatic(analyzer.cljc:3595)
	at cljs.analyzer$analyze_STAR_.invoke(analyzer.cljc:3586)
	at cljs.analyzer$analyze.invokeStatic(analyzer.cljc:3616)
	at cljs.analyzer$analyze.invoke(analyzer.cljc:3598)
	at cljs.analyzer$analyze.invokeStatic(analyzer.cljc:3607)
	at cljs.analyzer$analyze.invoke(analyzer.cljc:3598)
	at cljs.analyzer$analyze.invokeStatic(analyzer.cljc:3605)
	at cljs.analyzer$analyze.invoke(analyzer.cljc:3598)
	at cljs.analyzer$analyze_js_star_STAR_$fn__2299.invoke(analyzer.cljc:3029)
	at clojure.core$map$fn__5587.invoke(core.clj:2747)
	at clojure.lang.LazySeq.sval(LazySeq.java:40)
	at clojure.lang.LazySeq.seq(LazySeq.java:49)
	at clojure.lang.Cons.next(Cons.java:39)
	at clojure.lang.PersistentVector.create(PersistentVector.java:73)
	at clojure.lang.LazilyPersistentVector.create(LazilyPersistentVector.java:44)
	at clojure.core$vec.invokeStatic(core.clj:377)
	at clojure.core$vec.invoke(core.clj:367)
	at cljs.analyzer$analyze_js_star_STAR_.invokeStatic(analyzer.cljc:3029)
	at cljs.analyzer$analyze_js_star_STAR_.invoke(analyzer.cljc:3027)
	at cljs.analyzer$analyze_js_star.invokeStatic(analyzer.cljc:3057)
	at cljs.analyzer$analyze_js_star.invoke(analyzer.cljc:3056)
	at cljs.analyzer$fn__2312.invokeStatic(analyzer.cljc:3064)
	at cljs.analyzer$fn__2312.invoke(analyzer.cljc:3059)
	at clojure.lang.MultiFn.invoke(MultiFn.java:251)
	at cljs.analyzer$analyze_seq_STAR_.invokeStatic(analyzer.cljc:3348)
	at cljs.analyzer$analyze_seq_STAR_.invoke(analyzer.cljc:3346)
	at cljs.analyzer$analyze_seq_STAR__wrap.invokeStatic(analyzer.cljc:3353)
	at cljs.analyzer$analyze_seq_STAR__wrap.invoke(analyzer.cljc:3351)
	at cljs.analyzer$analyze_seq.invokeStatic(analyzer.cljc:3377)
	at cljs.analyzer$analyze_seq.invoke(analyzer.cljc:3355)
	at cljs.analyzer$analyze_form.invokeStatic(analyzer.cljc:3545)
	at cljs.analyzer$analyze_form.invoke(analyzer.cljc:3541)
	at cljs.analyzer$analyze_STAR_.invokeStatic(analyzer.cljc:3595)
	at cljs.analyzer$analyze_STAR_.invoke(analyzer.cljc:3586)
	at cljs.analyzer$analyze.invokeStatic(analyzer.cljc:3616)
	at cljs.analyzer$analyze.invoke(analyzer.cljc:3598)
	at cljs.analyzer$analyze_seq.invokeStatic(analyzer.cljc:3378)
	at cljs.analyzer$analyze_seq.invoke(analyzer.cljc:3355)
	at cljs.analyzer$analyze_form.invokeStatic(analyzer.cljc:3545)
	at cljs.analyzer$analyze_form.invoke(analyzer.cljc:3541)
	at cljs.analyzer$analyze_STAR_.invokeStatic(analyzer.cljc:3595)
	at cljs.analyzer$analyze_STAR_.invoke(analyzer.cljc:3586)
	at cljs.analyzer$analyze.invokeStatic(analyzer.cljc:3616)
	at cljs.analyzer$analyze.invoke(analyzer.cljc:3598)
	at cljs.analyzer$analyze.invokeStatic(analyzer.cljc:3607)
	at cljs.analyzer$analyze.invoke(analyzer.cljc:3598)
	at cljs.analyzer$fn__1733$fn__1736.invoke(analyzer.cljc:1541)
	at cljs.analyzer$fn__1733.invokeStatic(analyzer.cljc:1539)
	at cljs.analyzer$fn__1733.invoke(analyzer.cljc:1473)
	at clojure.lang.MultiFn.invoke(MultiFn.java:251)
	at cljs.analyzer$analyze_seq_STAR_.invokeStatic(analyzer.cljc:3348)
	at cljs.analyzer$analyze_seq_STAR_.invoke(analyzer.cljc:3346)
	at cljs.analyzer$analyze_seq_STAR__wrap.invokeStatic(analyzer.cljc:3353)
	at cljs.analyzer$analyze_seq_STAR__wrap.invoke(analyzer.cljc:3351)
	at cljs.analyzer$analyze_seq.invokeStatic(analyzer.cljc:3377)
	at cljs.analyzer$analyze_seq.invoke(analyzer.cljc:3355)
	at cljs.analyzer$analyze_form.invokeStatic(analyzer.cljc:3545)
	at cljs.analyzer$analyze_form.invoke(analyzer.cljc:3541)
	at cljs.analyzer$analyze_STAR_.invokeStatic(analyzer.cljc:3595)
	at cljs.analyzer$analyze_STAR_.invoke(analyzer.cljc:3586)
	at cljs.analyzer$analyze.invokeStatic(analyzer.cljc:3616)
	at cljs.analyzer$analyze.invoke(analyzer.cljc:3598)
	at cljs.compiler$emit_source.invokeStatic(compiler.cljc:1386)
	at cljs.compiler$emit_source.invoke(compiler.cljc:1365)
	at cljs.compiler$compile_file_STAR_$fn__3672.invoke(compiler.cljc:1467)
	at cljs.compiler$with_core_cljs.invokeStatic(compiler.cljc:1285)
	at cljs.compiler$with_core_cljs.invoke(compiler.cljc:1274)
	at cljs.compiler$compile_file_STAR_.invokeStatic(compiler.cljc:1451)
	at cljs.compiler$compile_file_STAR_.invoke(compiler.cljc:1444)
	at cljs.compiler$compile_file$fn__3702.invoke(compiler.cljc:1547)
	... 58 more
Caused by: java.io.FileNotFoundException: node_modules/web3/dist/web3.min.js (No such file or directory)
	at java.io.FileInputStream.open0(Native Method)
	at java.io.FileInputStream.open(FileInputStream.java:195)
	at java.io.FileInputStream.<init>(FileInputStream.java:138)
	at clojure.java.io$fn__10968.invokeStatic(io.clj:229)
	at clojure.java.io$fn__10968.invoke(io.clj:229)
	at clojure.java.io$fn__10881$G__10874__10888.invoke(io.clj:69)
	at clojure.java.io$fn__10980.invokeStatic(io.clj:258)
	at clojure.java.io$fn__10980.invoke(io.clj:254)
	at clojure.java.io$fn__10881$G__10874__10888.invoke(io.clj:69)
	at clojure.java.io$fn__10942.invokeStatic(io.clj:165)
	at clojure.java.io$fn__10942.invoke(io.clj:165)
	at clojure.java.io$fn__10894$G__10870__10901.invoke(io.clj:69)
	at clojure.java.io$reader.invokeStatic(io.clj:102)
	at clojure.java.io$reader.doInvoke(io.clj:86)
	at clojure.lang.RestFn.invoke(RestFn.java:410)
	at clojure.lang.AFn.applyToHelper(AFn.java:154)
	at clojure.lang.RestFn.applyTo(RestFn.java:132)
	at clojure.core$apply.invokeStatic(core.clj:659)
	at clojure.core$slurp.invokeStatic(core.clj:6862)
	at clojure.core$slurp.doInvoke(core.clj:6862)
	at clojure.lang.RestFn.invoke(RestFn.java:410)
	at status_im.utils.slurp$slurp.invokeStatic(slurp.clj:6)
	at status_im.utils.slurp$slurp.invoke(slurp.clj:5)
	at clojure.lang.AFn.applyToHelper(AFn.java:160)
	at clojure.lang.AFn.applyTo(AFn.java:144)
	at clojure.core$apply.invokeStatic(core.clj:661)
	at clojure.core$apply.invoke(core.clj:652)
	at cljs.analyzer$macroexpand_1_STAR_$fn__2369.invoke(analyzer.cljc:3295)
	at cljs.analyzer$macroexpand_1_STAR_.invokeStatic(analyzer.cljc:3294)
	at cljs.analyzer$macroexpand_1_STAR_.invoke(analyzer.cljc:3281)
	... 134 more
Error encountered performing task 'cljsbuild' with profile(s): 'prod'
Subprocess failed
make[1]: *** [Makefile:95: prod-build-android] Error 1
make[1]: Leaving directory '/home/hlolli/forks/status-react'
make: *** [Makefile:67: release-android] Error 2
```

Givan that nix-patch-realm is the same on my fork and on status-im (I'm checked out on my fork)

hlolli commented:
Btw, maybe you want to have a look at my clj2nix library which I used here for lumo https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/interpreters/clojurescript/lumo/deps.nix there's a function that generates a classpath, which can be passed to leiningen, tough I'd recomment tools.deps.alpha.

PombeirP commented:
Ah no, it was the lack of `--bypass-cache` in mobile. I'll push that, maybe you can reset and force push again? Sorry about that.

Thanks for the awesome help!

PombeirP commented:
> Btw, maybe you want to have a look at my clj2nix library which I used here for lumo https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/interpreters/clojurescript/lumo/deps.nix there's a function that generates a classpath, which can be passed to leiningen, tough I'd recomment tools.deps.alpha.

Yeah, that's one of the next steps after I bring yarn2nix to the project.

PombeirP commented:
@hlolli I pushed some more changes (added comments, fixed node version string in .TOOLVERSIONS).

hlolli commented:
Cool, force pushed, I'll have to pass for testing for today, my hands are very full for the rest of the evening.

PombeirP commented:
@hlolli Come to think about it, I believe that ubuntu-server.js is only used in desktop configuration, so we could leave the mobile part out of this for the time being (it'll be useful soon for https://github.com/status-im/status-react/issues/8203 though). Maybe we can leave the folder in there but just not bring it in to the mobile configuration.
